src/main/java/appeng/core/sync/packets/PacketInventoryAction.java

```diff
@@ -107,66 +78,106 @@ public PacketInventoryAction( final InventoryAction action, final int slot, fina
-	public void serverPacketData( final INetworkInfo manager, final AppEngPacket packet, final EntityPlayer player )
+	public AppEngPacket onMessage( final PacketInventoryAction message, final MessageContext ctx )
 	{
-		final EntityPlayerMP sender = (EntityPlayerMP) player;
-		if( sender.openContainer instanceof AEBaseContainer )
+		if( ctx.side == Side.CLIENT )
 		{
-			final AEBaseContainer baseContainer = (AEBaseContainer) sender.openContainer;
-			if( this.action == InventoryAction.AUTO_CRAFT )
+			if( message.action == InventoryAction.UPDATE_HAND )
 			{
-				final ContainerOpenContext context = baseContainer.openContext;
-				if( context != null )
+				if( message.slotItem == null )
 				{
-					final TileEntity te = context.getTile();
-					Platform.openGUI( sender, te, baseContainer.openContext.side, GuiBridge.GUI_CRAFTING_AMOUNT );
-
-					if( sender.openContainer instanceof ContainerCraftAmount )
+					ClientHelper.proxy.getPlayers().get( 0 ).inventory.setItemStack( null );
+				}
+				else
+				{
+					ClientHelper.proxy.getPlayers().get( 0 ).inventory.setItemStack( message.slotItem.getItemStack() );
+				}
+			}
+		}
+		else
+		{
+			final EntityPlayerMP sender = (EntityPlayerMP) ctx.getServerHandler().playerEntity;
+			if( sender.openContainer instanceof AEBaseContainer )
+			{
+				final AEBaseContainer baseContainer = (AEBaseContainer) sender.openContainer;
+				if( message.action == InventoryAction.AUTO_CRAFT )
+				{
+					final ContainerOpenContext context = baseContainer.openContext;
+					if( context != null )
 					{
-						final ContainerCraftAmount cca = (ContainerCraftAmount) sender.openContainer;
+						final TileEntity te = context.getTile();
+						Platform.openGUI( sender, te, baseContainer.openContext.side, GuiBridge.GUI_CRAFTING_AMOUNT );
 
-						if( baseContainer.getTargetStack() != null )
+						if( sender.openContainer instanceof ContainerCraftAmount )
 						{
-							cca.craftingItem.putStack( baseContainer.getTargetStack().getItemStack() );
-							cca.whatToMake = baseContainer.getTargetStack();
-						}
+							final ContainerCraftAmount cca = (ContainerCraftAmount) sender.openContainer;
+
+							if( baseContainer.getTargetStack() != null )
+							{
+								cca.craftingItem.putStack( baseContainer.getTargetStack().getItemStack() );
+								cca.whatToMake = baseContainer.getTargetStack();
+							}
 
-						cca.detectAndSendChanges();
+							cca.detectAndSendChanges();
+						}
 					}
 				}
+				else
+				{
+					baseContainer.doAction( sender, message.action, message.slot, message.id );
+				}
 			}
-			else
+		}
+
+		return null;
+	}
+
+	@Override
+	public void fromBytes( final ByteBuf buf )
+	{
+		this.action = InventoryAction.values()[buf.readInt()];
+		this.slot = buf.readInt();
+		this.id = buf.readLong();
+		final boolean hasItem = buf.readBoolean();
+		if( hasItem )
+		{
+			try
+			{
+				this.slotItem = AEItemStack.loadItemStackFromPacket( buf );
+			}
+			catch( final IOException e )
 			{
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Either remove or fill this block of code. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00108)